### ISSUE TITLE
[NEXT-1318] Unable to use context.waitUntil in Route Handler?

### ISSUE THREAD

**Issue by s1973** [vercel/next.js#50522]
### What is the improvement or update you wish to see?

When using pages router, we can use context on api routers:
```ts
import { NextResponse } from 'next/server';
import type { NextFetchEvent, NextRequest } from 'next/server';
 
export const config = {
  runtime: 'edge',
};
 
const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
 
async function getAlbum() {
  const res = await fetch('https://jsonplaceholder.typicode.com/albums/1');
  await wait(10000);
  return res.json();
}
 
export default function MyEdgeFunction(
  request: NextRequest,
  context: NextFetchEvent,
) {
  context.waitUntil(getAlbum().then((json) => console.log({ json })));
 
  return NextResponse.json({
    name: `Hello, from ${request.url} I'm an Edge Function!`,
  });
}
```
while using app router, we got this from [doc](https://nextjs.org/docs/app/api-reference/file-conventions/route#context-optional):

> Currently, the only value of context is params, which is an object containing the [dynamic route parameters](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) for the current route.

so I am wondering how to use NextFetchEvent with route handler?

### Is there any context that might help us understand?

I am trying to keep the function running after a response has been sent, as the vercel [doc](https://vercel.com/docs/concepts/functions/edge-functions/edge-functions-api#waituntil) navigated, we can use `waitUntil()`, but it seems not approachable in App Router and Route Handler, can I get any thoughts? thanks

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/app/api-reference/file-conventions/route#context-optional

<sub>[NEXT-1318](https://linear.app/vercel/issue/NEXT-1318/unable-to-use-contextwaituntil-in-route-handler)</sub>

**Comment by altechzilla** [comment]
Hey @s1973 have you managed to figure this out yet?

**Comment by altechzilla** [comment]
@balazsorban44 is this a Next.js bug? And if so, any estimate as to when context.waitUntil would become available?

**Comment by balazsorban44** [comment]
Hi, it's more like a missing feature. when Route Handlers have been introduced this was not added to the initial implementation requirements.

But after some internal discussions, we think we should add it so there's feature parity with API Routes (edge).

It's now tracked internally.

**Comment by altechzilla** [comment]
> Hi, it's more like a missing feature. when Route Handlers have been introduced this was not added to the initial implementation requirements.
> 
> But after some internal discussions, we think we should add it so there's feature parity with API Routes (edge).
> 
> It's now tracked internally.

Got it. Thank you! Will keep an eye out since it's blocking us from migrating some pretty important routes to the app directory.

**Comment by altechzilla** [comment]
Hey @balazsorban44 , is this being worked on yet by any chance? If so can I have a rough ETA? Thank you & I really appreciate it!

**Comment by sauhaardac** [comment]
+1 on this

Kind of strange that this was released without this feature. The types are literally wrong in saying there is this property that doesn't exist (waitUntil). Had to move some routes to page router as a result of it.

**Comment by anuraagvaidya** [comment]
> +1 on this
> 
> Kind of strange that this was released without this feature. The types are literally wrong in saying there is this property that doesn't exist (waitUntil). Had to move some routes to page router as a result of it.

Were you able to access the context object using page router? Mine is an empty object.

**Comment by anuraagvaidya** [comment]
I was able to get around this issue by moving my long-running functions to `middleware.js`

`export async function middleware(req, context) {
      //if() condition to check the path
     //{ use context.waitUntil() }
}`

**Comment by altechzilla** [comment]
Great workaround @anuraagvaidya , but pretty inconvenient for large apps with lots of moving pieces & routes. Really hoping to see native context support at a route level

**Comment by sam3d** [comment]
Just to add that it would also be really helpful to be able to execute `context.waitUntil` in a server action

**Comment by matannahmani** [comment]
any updates on this?

**Comment by sam3d** [comment]
I've discovered a workaround to do this on edge functions with both route handlers and server actions.

https://github.com/vercel/next.js/blob/445e70502834540d476b8eeaed0228241acd92eb/packages/next/src/server/web/internal-edge-wait-until.ts#L36

You can use it like so:

```ts
"use server";

import { internal_runWithWaitUntil } from "next/dist/server/web/internal-edge-wait-until";

export async function action() {
  console.log("Request received");
  internal_runWithWaitUntil(async () => {
    await new Promise((resolve) => setTimeout(resolve, 10000));
    console.log("Waited for 10 seconds")
  });
  return true;
}
```

This is pretty clearly a highly experimental and unstable API, and I'm not entirely sure what impact using it like this could have (so use it at your own risk). That said, I've been using it on Vercel for a couple of days and I haven't encountered any issues so far.

**Comment by sam3d** [comment]
Also, I couldn't find this anywhere in the Vercel docs but it's interesting nonetheless. Considering that Cloudflare Workers area what's under the hood of Edge Functions, then this is the timeout limit of `waitUntil`:

> Duration is a measurement of wall-clock time — the total amount of time from the start to end of an invocation of a Worker. There is no hard limit on the duration of a Worker. As long as the client that sent the request remains connected, the Worker can continue processing, making subrequests, and setting timeouts on behalf of that request. When the client disconnects, all tasks associated with that client request are canceled. **Use [event.waitUntil()](https://developers.cloudflare.com/workers/runtime-apis/handlers/fetch/) to delay cancellation for another 30 seconds or until the promise passed to waitUntil() completes.**

https://developers.cloudflare.com/workers/platform/limits/

**Comment by maccman** [comment]
@sam3d I'm wondering if there's an update on when this is being planned to be released?

**Comment by PSoltes** [comment]
Yeah feels weird that there were 2(!!) major versions with app router released without this kinda important feature.

**Comment by sam3d** [comment]
> @sam3d I'm wondering if there's an update on when this is being planned to be released?

Idk lol I just found it 🤷‍♀️

That said, I've been using it in prod for months and it's never faltered. I think the API is unstable (like `unstable_cache`), but the implementation is really solid.

**Comment by kyb3r** [comment]
using internal_runWithWaitUntil

But any eta when the api will be stabilised in a release?

**Comment by maccman** [comment]
It's brewing! https://github.com/vercel/next.js/pull/65038

**Comment by Thinkscape** [comment]
Good news!

https://vercel.com/changelog/waituntil-is-now-available-for-vercel-functions
https://vercel.com/docs/functions/functions-api-reference#waituntil

**Comment by mmkal** [comment]
It sounds like the above is specific to functions? It's not clear to me if it's supposed to be used in:

- edge middleware?
- node middleware?
- edge route handlers?
- node route handlers?
- edge server components?
- node server components?

I had a look at the implementation and it looks like it's getting context from `globalThis`, but it null checks both the context's existence and the `waitUntil` property on the context.

<img width="670" alt="image" src="https://github.com/user-attachments/assets/2a9f5289-a5d4-4ddb-a200-e84171be500f">

And from adding some logs, it seems for me, while running locally, I haven't observed that context to exist, so `waitUntil` will silently do nothing at all.

Maybe the context will exist when deployed, and it will do something - but it doesn't inspire confidence, because if it doesn't exist. And probably most of the time, your promise will be resolved anyway, since you've already kicked it off, and maybe sometimes it resolves fast enough, before your container shuts down.

It's also not clear to me how to verify that it's working because I'm using it correctly, or by blind luck.

Could it be that the happy users of `waitUntil` are just relying on floating promises working *some* of the time? Am I missing something?

**Comment by lubieowoce** [comment]
> I'm still not sure why the docs only reference functions when it seems to work everywhere

It's just our terminology for "a serverless thing deployed on Vercel". this includes next.js apps and thus pages, actions, route handlers, and middleware. It also works in both node and edge runtimes.

> Not sure about non-vercel deployment environments

`unstable_after(promise)` is equivalent to `waitUntil`, and is intended as the provider-agnostic way of doing this. But that generally needs some support from the hosting provider, at least on serverless hosts -- in that kind of setup we can't do `waitUntil`-like behavior without platform integration.

**Comment by marcoshernanz** [comment]
Thanks for reporting this. The underlying use case is now supported by [`after()`](https://nextjs.org/docs/app/api-reference/functions/after) from `next/server`, including Route Handlers in both Node.js and Edge runtimes. It was introduced as `unstable_after` in the Next.js 15 RC and stabilized in Next.js 15.1. `NextFetchEvent` is not part of the Route Handler context, so use `after(() => ...)` for post-response work. Closing as resolved.
